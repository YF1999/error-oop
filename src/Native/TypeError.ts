import { IErrorArguments } from './IErrorArguments';
import { appendInnerErrorStack, setNonEnumerable } from './Tools';
import { IErrorOptions } from './Types';

/**
 * Represents an error when a value is not of the expected type.
 * This is roughly the same as the native `TypeError` class. It additionally supports an innerError attribute.
 */
export class NativeTypeError extends TypeError {
    #innerError: Error | null;

    public constructor();
    /**
     * @param message The error message that explains the reason for this error.
     */
    public constructor(message: string);
    /**
     * @param message The error message that explains the reason for this error.
     * @param innerError The error that is the cause of the current error. Stack trace will be appended.
     */
    public constructor(message: string, innerError: Error);
    /**
     * @param options The constructor options.
     */
    public constructor(options: IErrorOptions);

    public constructor(...[arg1, arg2]: IErrorArguments<IErrorOptions>) {
        super(typeof arg1 === 'object' ? arg1.message : arg1);

        this.#innerError = (typeof arg1 === 'object' ? arg1.innerError : arg2) ?? null;

        // When the first call to `stack` property happens, it will combine `name` and `message` with trace stack to
        // `stack` property, we should generate message before this call.
        this.stack = appendInnerErrorStack(this.stack, this.#innerError);

        this._setNonEnumerable('stack');
    }

    public get name(): string {
        return this.constructor.name;
    }

    public get innerError(): Error | null {
        return this.#innerError;
    }

    protected _setNonEnumerable(property: string): void {
        setNonEnumerable(this, property);
    }
}

export { NativeTypeError as TypeError };
