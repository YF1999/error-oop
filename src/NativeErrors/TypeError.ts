import { ErrorOptions, TypeErrorMessageProps, TypeErrorProps } from '../CommonTypes';
import { appendInnerErrorStack, setNonEnumerable } from '../utils';

export abstract class AbstractTypeError extends TypeError {
    protected _innerError?: Error;

    public constructor(props: TypeErrorProps, options: ErrorOptions<TypeErrorMessageProps>) {
        super();

        const { message, innerError } = props;
        const { name, generateMessage } = options;

        this._innerError = innerError;
        this.name = name || this.constructor.name;
        this.message = generateMessage ? generateMessage({ name: this.name, message }) : message;

        // When the first call to `stack` property happens, it will combine `name` and `message`
        // with trace stack to `stack` property, we should generate message before this call.
        this.stack = appendInnerErrorStack(this.stack, this._innerError);

        this._setNonEnumerable('name');
        this._setNonEnumerable('message');
        this._setNonEnumerable('stack');
        this._setNonEnumerable('_innerError');
    }

    public get innerError(): Error | undefined {
        return this._innerError;
    }

    protected _setNonEnumerable(property: string): void {
        setNonEnumerable(this, property);
    }
}

/**
 * Represents an error when a value is not of the expected type. This is roughly the same as the
 * native TypeError class. It additionally supports an innerError attribute.
 */
export class TypeErrorPro extends AbstractTypeError {
    public constructor();
    /**
     * @param message The error message that explains the reason for this error.
     */
    public constructor(message: string);
    /**
     * @param message The error message that explains the reason for this error.
     * @param innerError The error that is the cause of the current error. Stack trace will be
     * appended.
     */
    public constructor(message: string, innerError: Error);

    public constructor(message: string = '', innerError?: Error) {
        super({ message, innerError }, { name: 'TypeError' });
    }
}

export { TypeErrorPro as TypeError };
