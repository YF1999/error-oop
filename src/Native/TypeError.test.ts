import { NativeTypeError } from './TypeError';

standardTestError(NativeTypeError, 'NativeTypeError', 'error');
