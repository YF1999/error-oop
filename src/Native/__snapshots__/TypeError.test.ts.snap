// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test \`NativeTypeError\` own keys 1`] = `
Array [
  "stack",
]
`;

exports[`Test \`NativeTypeError\` own keys 2`] = `
Array [
  "stack",
  "message",
]
`;

exports[`Test \`NativeTypeError\` own keys 3`] = `
Array [
  "stack",
  "message",
]
`;

exports[`Test \`NativeTypeError\` own keys 4`] = `
Array [
  "stack",
  "message",
]
`;
